[MINOR][SS][DOC] Added missing config `maxFileAge` in file streaming source

linehrr opened the pull request:
## What changes were proposed in this pull request?

added the missing config for structured streaming when using file source.

from the code we have 
```
  /**
   * Maximum age of a file that can be found in this directory, before it is ignored. For the
   * first batch all files will be considered valid. If `latestFirst` is set to `true` and
   * `maxFilesPerTrigger` is set, then this parameter will be ignored, because old files that are
   * valid, and should be processed, may be ignored. Please refer to SPARK-19813 for details.
   *
   * The max age is specified with respect to the timestamp of the latest file, and not the
   * timestamp of the current system. That this means if the last file has timestamp 1000, and the
   * current system time is 2000, and max age is 200, the system will purge files older than
   * 800 (rather than 1800) from the internal state.
   *
   * Default to a week.
   */
  val maxFileAgeMs: Long =
    Utils.timeStringAsMs(parameters.getOrElse("maxFileAge", "7d"))
```
which is not documented. 

also the file processing order was not mentioned but in the code we specifically select the file list based on file mtime: 
```scala
private val fileSortOrder = if (sourceOptions.latestFirst) {
      logWarning(
        """'latestFirst' is true. New files will be processed first, which may affect the watermark
          |value. In addition, 'maxFileAge' will be ignored.""".stripMargin)
      implicitly[Ordering[Long]].reverse
    } else {
      implicitly[Ordering[Long]]
    }

val files = allFiles.sortBy(_.getModificationTime)(fileSortOrder).map { status =>
      (status.getPath.toUri.toString, status.getModificationTime)
    }
```
---------
![Screen Shot 2019-05-07 at 5 55 01 PM](https://user-images.githubusercontent.com/1124115/57335683-5a8b0400-70f1-11e9-98c8-99f173872842.png)
---------
![Screen Shot 2019-05-07 at 5 54 55 PM](https://user-images.githubusercontent.com/1124115/57335684-5a8b0400-70f1-11e9-996a-4bb1639e3d6b.png)

Commit:
fixed integer overflow in rateLimit calculation

added UT to against the calculation, setting maxRatePerTrigger to Integer.MaxValue
 Changes to be committed:
	modified:   src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala
	modified:   src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
linting code

 On branch master
 Your branch is up to date with 'origin/master'.

 Changes to be committed:
	modified:   external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala
	modified:   external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala
	modified:   external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
adjust ticket number position in test desc

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
removed unnecessary type conversion

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
removed useless implicits imports

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
made test case faster

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
fixed comments

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
added missing doc for file streaming source

 On branch master
 Changes to be committed:
	modified:   docs/structured-streaming-programming-guide.md

AmplabJenkins commented:
Can one of the admins verify this patch?

Commit:
added default value in doc

 Changes to be committed:
	modified:   docs/structured-streaming-programming-guide.md

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
Hi, @linehrr (ryne.yang). Thank you for your second contribution. Am I right? 
BTW, for the doc change, it's recommended to generate the doc and attach the new screen shot into the PR description.

Commit:
added more desc on file processing order

 Changes to be committed:
	modified:   docs/structured-streaming-programming-guide.md

linehrr commented:
@dongjoon-hyun 
ok attached pics, edited 2 places regarding file source processing order and `maxFileAge` option.

dongjoon-hyun commented:
ok to test

dongjoon-hyun commented:
Thanks, @linehrr .

dongjoon-hyun commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -541,6 +541,8 @@ Here are the details of all the sources in Spark.
         <br/>
         <code>fileNameOnly</code>: whether to check new files based on only the filename instead of on the full path (default: false). With this set to `true`, the following files would be considered as the same file, because their filenames, "dataset.txt", are the same:
         <br/>
+        <code>maxFileAge</code>: Maximum age of a file that can be found in this directory, before it is ignored. For the first batch all files will be considered valid. If `latestFirst` is set to `true` and `maxFilesPerTrigger` is set, then this parameter will be ignored, because old files that are valid, and should be processed, may be ignored. The max age is specified with respect to the times ...
```
According to the attached image, it seems that we need the following changes.
```
- If `latestFirst` is set to `true` and `maxFilesPerTrigger`
+ If <code>latestFirst</code> is set to `true` and <code>maxFilesPerTrigger</code>
```

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10516/
Test PASSed.

SparkQA commented:
**[Test build #105234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105234/testReport)** for PR 24548 at commit [`4a0ec5c`](https://github.com/apache/spark/commit/4a0ec5c576a73143cf87544d9c00572ba9fe76e2).

SparkQA commented:
**[Test build #105234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105234/testReport)** for PR 24548 at commit [`4a0ec5c`](https://github.com/apache/spark/commit/4a0ec5c576a73143cf87544d9c00572ba9fe76e2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105234/
Test PASSed.

Commit:
fixed font

 Changes to be committed:
	modified:   structured-streaming-programming-guide.md

SparkQA commented:
**[Test build #105259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105259/testReport)** for PR 24548 at commit [`1a4a68a`](https://github.com/apache/spark/commit/1a4a68a28612210981797fd375761639bd38f32c).

linehrr commented:
Fixed the font.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10539/
Test PASSed.

SparkQA commented:
**[Test build #105259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105259/testReport)** for PR 24548 at commit [`1a4a68a`](https://github.com/apache/spark/commit/1a4a68a28612210981797fd375761639bd38f32c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105259/
Test PASSed.

srowen commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -510,7 +510,7 @@ returned by `SparkSession.readStream()`. In [R](api/R/read.stream.html), with th
 There are a few built-in sources.
 
-  - **File source** - Reads files written in a directory as a stream of data. Supported file formats are text, csv, json, orc, parquet. See the docs of the DataStreamReader interface for a more up-to-date list, and supported options for each file format. Note that the files must be atomically placed in the given directory, which in most file systems, can be achieved by file move operations.
+  - **File source** - Reads files written in a directory as a stream of data(files will be processed in the order of file's modification time, this is to mimic the time series feature from streaming nature. If `latestFirst` is set, order will be reversed). Supported file formats are text, csv, json, orc, parquet. See the docs of the DataStreamReader interface for a more up-to-date list, and suppo ...
```
I think the text need a number of fixes:

```
Reads files written in a directory as a stream of data. Files will be processed in the order of file modification time. If `latestFirst` is set, order will be reversed. Supported file formats are text, CSV, JSON, ORC, Parquet. See the docs of the DataStreamReader interface for a more up-to-date list, and supported options for each file format. Note that the files must be atomically placed in the given directory, which in most file systems, can be achieved by file move operations.
```

linehrr replied:
ok revised sentence.

Commit:
revised desc

 Changes to be committed:
	modified:   structured-streaming-programming-guide.md

SparkQA commented:
**[Test build #105266 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105266/testReport)** for PR 24548 at commit [`c10d65f`](https://github.com/apache/spark/commit/c10d65f6f17ab1a996e58c00bddd0b902bd64442).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10545/
Test PASSed.

SparkQA commented:
**[Test build #105266 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105266/testReport)** for PR 24548 at commit [`c10d65f`](https://github.com/apache/spark/commit/c10d65f6f17ab1a996e58c00bddd0b902bd64442).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105266/
Test PASSed.

felixcheung commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -510,8 +510,7 @@ returned by `SparkSession.readStream()`. In [R](api/R/read.stream.html), with th
 
-  - **File source** - Reads files written in a directory as a stream of data. Supported file formats are text, csv, json, orc, parquet. See the docs of the DataStreamReader interface for a more up-to-date list, and supported options for each file format. Note that the files must be atomically placed in the given directory, which in most file systems, can be achieved by file move operations.
-
+  - **File source** - Reads files written in a directory as a stream of data. Files will be processed in the order of file modification time. If `latestFirst` is set, order will be reversed. Supported file formats are text, CSV, JSON, ORC, Parquet. See the docs of the DataStreamReader interface for a more up-to-date list, and supported options for each file format. Note that the files must be ato ...
```
why are ` text, CSV, JSON, ORC, Parquet` capitalized? I thought file format names are lower cased

srowen replied:
Maybe identifiers in code like `csv` would be lowercase, but these formats are acronyms and are properly capitalized when named. CSV = Comma Separated Values, etc.

linehrr replied:
do we need to change this ?

srowen replied:
I don't believe so.

srowen commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -541,6 +540,8 @@ Here are the details of all the sources in Spark.
         <br/>
         <code>fileNameOnly</code>: whether to check new files based on only the filename instead of on the full path (default: false). With this set to `true`, the following files would be considered as the same file, because their filenames, "dataset.txt", are the same:
         <br/>
+        <code>maxFileAge</code>: Maximum age of a file that can be found in this directory, before it is ignored. For the first batch all files will be considered valid. If <code>latestFirst</code> is set to `true` and <code>maxFilesPerTrigger</code> is set, then this parameter will be ignored, because old files that are valid, and should be processed, may be ignored. The max age is specified wit ...
```
Not a big deal but can you fix the end here, actually: "system. (Default: 1 week)"

linehrr replied:
solved.

linehrr committed:
minor change

SparkQA commented:
**[Test build #105324 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105324/testReport)** for PR 24548 at commit [`a127f4c`](https://github.com/apache/spark/commit/a127f4cead1f85d2c0753779bdb9787c03e5fd21).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10596/
Test PASSed.

SparkQA commented:
**[Test build #105324 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105324/testReport)** for PR 24548 at commit [`a127f4c`](https://github.com/apache/spark/commit/a127f4cead1f85d2c0753779bdb9787c03e5fd21).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105324/
Test PASSed.

dongjoon-hyun commented:
Merged to master/branch-2.4.
